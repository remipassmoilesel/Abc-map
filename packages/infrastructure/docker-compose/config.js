module.exports = {
  environmentName: 'production',
  externalUrl: 'http://localhost:10082/',
  server: {
    host: '0.0.0.0',
    port: 10082,
    log: {
      requests: true,
      errors: true,
      warnings: false,
    },
    globalRateLimit: {
      max: 1000,
      timeWindow: '30m',
    },
    authenticationRateLimit: {
      max: 200,
      timeWindow: '30m',
    },
  },
  project: {
    maxPerUser: 10,
  },
  database: {
    url: 'mongodb://abc-mongodb:27017',
    username: 'CHANGEME',
    password: 'CHANGEME',
  },
  jwt: {
    algorithm: 'HS512',
  },
  authentication: {
    secret: 'CHANGEME',
    tokenExpiresIn: '7d',
    passwordLostExpiresIn: '30min',
  },
  registration: {
    passwordSalt: 'CHANGEME',
    secret: 'CHANGEME',
    confirmationExpiresIn: '24h',
  },
  smtp: {
    from: 'no-reply@abc-map.fr',
    host: 'abc-smtp',
    port: 25,
  },
  datastore: {
    path: '/srv/datastore',
  },
  legalMentions: `
            <div class="p-2">
                <h1>CHANGEME</h1>
            </div>
            `,
};
