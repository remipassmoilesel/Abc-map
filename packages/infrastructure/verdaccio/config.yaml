# Copyright © 2021 Rémi Pace.
# This file is part of Abc-Map.
#
# Abc-Map is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of
# the License, or (at your option) any later version.
#
# Abc-Map is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General
# Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.

storage: /tmp/verdaccio-storage
# We must use "127.0.0.1" and not localhost here, otherwise start may fail on some OS
listen: 127.0.0.1:4873
max_body_size: 50mb
auth:
  htpasswd:
    file: ./htpasswd
uplinks:
  npmjs:
    url: https://registry.npmjs.org/
packages:
  '@abc-map/*':
    access: $all
    publish: $all
  '**':
    access: $all
    publish: $all
    proxy: npmjs
middlewares:
  audit:
    enabled: false
logs:
  - { type: stdout, format: pretty, level: http }
