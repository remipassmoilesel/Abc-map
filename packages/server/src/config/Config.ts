/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { Algorithm } from 'jsonwebtoken';

export const LOCAL_ENVIRONMENT = 'local';
export const TEST_ENVIRONMENT = 'test';
export const STAGING_ENVIRONMENT = 'staging';

export interface ConfigInput {
  environmentName: string;
  development?: DevelopmentDataConfig;
  // External URL is the url that will be used by users. It must contain protocol and full domain. E.g: 'https://abc-map.fr'.
  externalUrl: string;
  server: ServerConfig;
  project: ProjectConfig;
  database: DatabaseConfig;
  jwt: JwtConfig;
  authentication: AuthenticationConfig;
  registration: RegistrationConfig;
  smtp: SmtpConfig;
  datastore: DatastoreConfig;
  legalMentions: string;
  webapp?: WebappConfig;
  userDocumentation?: UserDocConfig;
}

/**
 * This config is used in application
 */
export interface Config extends ConfigInput {
  webappPath: string;
  userDocumentationPath: string;
}

export interface DevelopmentDataConfig {
  // If set, data will be generated in order to facilitate development (users, projects, ...)
  generateData?: {
    users: number;
    projectsPerUser: number;
  };
  // If true, mails will not be sent but persisted on disk
  persistEmails?: boolean;
}

export interface ServerConfig {
  host: string;
  port: number;
  log: {
    requests: boolean;
    errors: boolean;
    warnings: boolean;
  };
  globalRateLimit: {
    max: number;
    timeWindow: string;
  };
  authenticationRateLimit: {
    max: number;
    timeWindow: string;
  };
}

export interface ProjectConfig {
  maxPerUser: number;
}

export interface DatabaseConfig {
  url: string;
  databaseName?: string;
  username: string;
  password: string;
}

export interface JwtConfig {
  algorithm: Algorithm;
}

export interface AuthenticationConfig {
  secret: string;
  tokenExpiresIn: string;
  passwordLostExpiresIn: string;
}

export interface RegistrationConfig {
  passwordSalt: string;
  secret: string;
  confirmationExpiresIn: string;
}

export interface SmtpConfig {
  from: string;
  host: string;
  port: number;
  auth?: {
    user: string;
    pass: string;
  };
}

export interface DatastoreConfig {
  path: string;
}

export interface WebappConfig {
  appendToBody?: string;
}

export interface UserDocConfig {
  appendToBody?: string;
}
