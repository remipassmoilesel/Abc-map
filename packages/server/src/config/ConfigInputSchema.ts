/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */
import { JSONSchemaType } from 'ajv';
import { ConfigInput } from './Config';

export const ConfigInputSchema: JSONSchemaType<ConfigInput> = {
  type: 'object',
  required: ['environmentName', 'externalUrl', 'server', 'project', 'database', 'jwt', 'authentication', 'registration', 'smtp', 'datastore', 'legalMentions'],
  additionalProperties: false,
  properties: {
    environmentName: { type: 'string' },
    externalUrl: { type: 'string' },
    server: {
      type: 'object',
      required: ['host', 'port', 'log', 'globalRateLimit', 'authenticationRateLimit'],
      properties: {
        host: { type: 'string' },
        port: { type: 'number' },
        log: {
          type: 'object',
          required: ['requests', 'errors', 'warnings'],
          properties: {
            requests: { type: 'boolean' },
            errors: { type: 'boolean' },
            warnings: { type: 'boolean' },
          },
        },
        globalRateLimit: {
          type: 'object',
          required: ['max', 'timeWindow'],
          properties: {
            max: { type: 'number' },
            timeWindow: { type: 'string' },
          },
        },
        authenticationRateLimit: {
          type: 'object',
          required: ['max', 'timeWindow'],
          properties: {
            max: { type: 'number' },
            timeWindow: { type: 'string' },
          },
        },
      },
    },
    project: {
      type: 'object',
      required: ['maxPerUser'],
      properties: {
        maxPerUser: { type: 'number' },
      },
    },
    database: {
      type: 'object',
      required: ['url', 'username', 'password'],
      properties: {
        url: { type: 'string' },
        databaseName: { type: 'string', nullable: true },
        username: { type: 'string' },
        password: { type: 'string' },
      },
    },
    jwt: {
      type: 'object',
      required: ['algorithm'],
      properties: {
        algorithm: { type: 'string' },
      },
    },
    authentication: {
      type: 'object',
      required: ['secret', 'tokenExpiresIn', 'passwordLostExpiresIn'],
      properties: {
        secret: { type: 'string' },
        tokenExpiresIn: { type: 'string' },
        passwordLostExpiresIn: { type: 'string' },
      },
    },
    registration: {
      type: 'object',
      required: ['passwordSalt', 'secret', 'confirmationExpiresIn'],
      properties: {
        passwordSalt: { type: 'string' },
        secret: { type: 'string' },
        confirmationExpiresIn: { type: 'string' },
      },
    },
    smtp: {
      type: 'object',
      required: ['from', 'host', 'port'],
      additionalProperties: false,
      properties: {
        from: { type: 'string' },
        host: { type: 'string' },
        port: { type: 'number' },
        auth: {
          type: 'object',
          required: [],
          nullable: true,
          properties: {
            user: { type: 'string' },
            pass: { type: 'string' },
          },
        },
      },
    },
    datastore: {
      type: 'object',
      required: ['path'],
      properties: {
        path: { type: 'string' },
      },
    },
    development: {
      type: 'object',
      required: [],
      nullable: true,
      properties: {
        generateData: {
          type: 'object',
          nullable: true,
          required: ['users'],
          properties: {
            users: { type: 'number' },
            projectsPerUser: { type: 'number' },
          },
        },
        persistEmails: {
          type: 'boolean',
          nullable: true,
        },
      },
    },
    legalMentions: {
      type: 'string',
    },
    webapp: {
      type: 'object',
      nullable: true,
      properties: {
        appendToBody: {
          type: 'string',
          nullable: true,
        },
      },
    },
    userDocumentation: {
      type: 'object',
      nullable: true,
      properties: {
        appendToBody: {
          type: 'string',
          nullable: true,
        },
      },
    },
  },
};
