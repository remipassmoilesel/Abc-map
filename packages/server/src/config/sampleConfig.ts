/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 *
 *
 *
 */

import { ConfigInput } from './Config';

/**
 * This function return a valid javascript configuration example.
 */
export function sampleConfig(): string {
  const config: ConfigInput = {
    environmentName: 'local',
    externalUrl: 'http://localhost:10082/',
    server: {
      host: 'localhost',
      port: 10_082,
      log: {
        requests: true,
        errors: true,
        warnings: false,
      },
      globalRateLimit: {
        max: 1000,
        timeWindow: '1m',
      },
      authenticationRateLimit: {
        max: 1000,
        timeWindow: '1m',
      },
    },
    project: {
      maxPerUser: 10,
    },
    database: {
      url: 'mongodb://localhost:27019',
      username: 'admin',
      password: 'admin',
    },
    jwt: {
      algorithm: 'HS512',
    },
    authentication: {
      secret: 'azerty1234',
      tokenExpiresIn: '6h',
      passwordLostExpiresIn: '30min',
    },
    registration: {
      passwordSalt: 'azerty1234',
      secret: 'azerty1234',
      confirmationExpiresIn: '24h',
    },
    smtp: {
      from: 'no-reply@abc-map.fr',
      host: 'smtp.ethereal.email',
      port: 587,
      auth: {
        user: 'lelia16@ethereal.email',
        pass: '63rntn3G4DU3uue2MJ',
      },
    },
    datastore: {
      path: 'resources/dev-datastore',
    },
    development: {
      generateData: {
        users: 100,
        projectsPerUser: 5,
      },
      persistEmails: true,
    },
    webapp: {
      appendToBody: '<div>Server templated content (webapp)</div>',
    },
    userDocumentation: {
      appendToBody: '<div>Server templated content (userDocumentation)</div>',
    },
    legalMentions: `<div>This platform is offered in order to be useful, but ....</div>`,
  };

  return `
module.exports = ${JSON.stringify(config, null, 2)}
`;
}
