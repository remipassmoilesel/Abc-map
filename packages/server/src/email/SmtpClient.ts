/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { Config } from '../config/Config';
import * as nodemailer from 'nodemailer';
import { Transporter } from 'nodemailer';
import { Logger } from '@abc-map/shared';
import { promises as fs } from 'fs';
import { constants } from 'fs';
import * as path from 'path';

const logger = Logger.get('SmtpClient.ts', 'info');

export function disableSmtpClientLogging() {
  logger.disable();
}

export class SmtpClient {
  private transporter: Transporter;

  constructor(private config: Config) {
    this.transporter = nodemailer.createTransport({
      from: config.smtp.from,
      host: config.smtp.host,
      port: config.smtp.port,
      auth: config.smtp.auth,
    });
  }

  public async sendMail(to: string, subject: string, body: string): Promise<void> {
    logger.info(`Sending mail to ${to}`);
    const html = this.htmlTemplate(subject, body);

    if (!this.config.development?.persistEmails) {
      await this.transporter.sendMail({
        from: this.config.smtp.from,
        to,
        subject,
        html,
      });
    } else {
      await this.fakeSending(to, html);
    }
  }

  private htmlTemplate(subject: string, body: string): string {
    /* eslint-disable max-len */
    return `
      <!DOCTYPE html>
      <html lang="en">
      <head>
        <meta charset="UTF-8">
        <title>${subject}</title>
        <style>
            body {
                font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif;";
            }
        </style>
      </head>
      <body>
          ${body}
      </body>
      </html>
    `;
    /* eslint-enable max-len */
  }

  private async fakeSending(email: string, body: string): Promise<void> {
    const mailDir = path.resolve(__dirname, '..', '..', '..', 'e2e-tests', 'emails');
    const mailPath = path.resolve(mailDir, `${email}.html`);
    logger.warn(`Mail will not be sent, you can see it at: ${mailPath}`);

    const dirExists = await fs
      .access(mailDir, constants.F_OK)
      .then(() => true)
      .catch(() => false);

    if (!dirExists) {
      await fs.mkdir(mailDir);
    }
    await fs.writeFile(mailPath, body);
  }
}
