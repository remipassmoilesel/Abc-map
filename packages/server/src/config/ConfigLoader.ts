/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { Env, EnvKey } from './Env';
import { Config, ConfigInput } from './Config';
import { errorMessage, Logger } from '@abc-map/shared';
import * as path from 'path';
import * as _ from 'lodash';
import { Validation } from '../utils/Validation';
import { isDirectory } from '../utils/isDirectory';
import { sampleConfig } from './sampleConfig';

export const logger = Logger.get('ConfigLoader.ts', 'info');

export type LoadConfigFunc = () => Promise<Config>;

export class ConfigLoader {
  public static readonly DEFAULT_CONFIG = 'resources/configuration/development.js';

  public static getPathFromEnv(): string {
    const env = new Env();
    return env.get(EnvKey.CONFIG) || ConfigLoader.DEFAULT_CONFIG;
  }

  public static async load(): Promise<Config> {
    const configPath = ConfigLoader.getPathFromEnv();
    return new ConfigLoader().load(configPath);
  }

  public static safeConfig(config: Config): Config {
    const safe: Config = _.cloneDeep(config);
    const safeValue = '<value-masked>';
    safe.authentication.secret = safeValue;
    safe.database.password = safeValue;
    safe.registration.passwordSalt = safeValue;
    safe.registration.secret = safeValue;
    if (safe.smtp.auth?.pass) {
      safe.smtp.auth.pass = safeValue;
    }

    safe.legalMentions = `${safe.legalMentions.substring(0, 50)}...`;

    return safe;
  }

  public async load(configPath: string): Promise<Config> {
    const _configPath = path.resolve(configPath);

    const explainedError = (message: string) => {
      logger.error(
        [
          '',
          '',
          '⚠️   Invalid configuration. ',
          '',
          'Message: ' + message,
          '',
          'This is a valid configuration example:',
          '',
          '------ config.js ------',
          sampleConfig()
            .split('\n')
            .map((line) => '   ' + line)
            .join('\n'),
          '------ config.js ------',
          '',
          'You can see other examples here: https://gitlab.com/abc-map/abc-map/-/tree/master/packages/server/resources/configuration',
          '',
          '',
        ].join('\n')
      );
    };

    let input: ConfigInput;
    try {
      /* eslint-disable @typescript-eslint/no-var-requires */
      input = _.cloneDeep(require(_configPath));
    } catch (err) {
      explainedError(errorMessage(err));
      return Promise.reject(new Error(`Cannot load configuration ${configPath}: ${errorMessage(err)}`));
    }

    // We validate config input
    if (!Validation.ConfigInput(input)) {
      explainedError(Validation.formatErrors(Validation.ConfigInput));
      return Promise.reject(new Error(`Configuration ${configPath} is not valid: ${Validation.formatErrors(Validation.ConfigInput)}`));
    }

    const config: Config = {
      ...input,
      webappPath: path.resolve(__dirname, '../../public/webapp'),
      userDocumentationPath: path.resolve(__dirname, '../../public/user-documentation'),
    };

    // We check if datastore is a directory
    if (!(await isDirectory(config.datastore.path))) {
      throw new Error(`Datastore root '${config.datastore.path}' must be a directory`);
    }

    // We remove trailing slash if present
    config.externalUrl = config.externalUrl.trim();
    if (config.externalUrl.slice(-1) === '/') {
      config.externalUrl = config.externalUrl.slice(0, -1);
    }

    // We check if webapp path is a directory
    if (!(await isDirectory(config.webappPath))) {
      throw new Error(`Webapp root '${config.webappPath}' must be a directory. Try "abc build" on a developer workstation.`);
    }

    return config;
  }
}
