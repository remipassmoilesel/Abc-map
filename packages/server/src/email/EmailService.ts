/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbstractService } from '../services/AbstractService';
import { SmtpClient } from './SmtpClient';
import { Config } from '../config/Config';
import { WebappRoutes, Language, ConfirmAccountTokenParam, ResetPasswordTokenParam } from '@abc-map/shared';

const Routes = {
  [Language.French]: new WebappRoutes(Language.French),
  [Language.English]: new WebappRoutes(Language.English),
};

const footerFr = (config: Config) => `
  <p>&nbsp;</p>
  <small>Ceci est un message automatique, envoyé par la plateforme <a href="${config.externalUrl}">${config.externalUrl}</a>.
  Vous ne pouvez pas répondre à ce message.</small>
`;

const footerEn = (config: Config) => `
  <p>&nbsp;</p>
  <small>This is an automatic message, sent by the platform <a href="${config.externalUrl}">${config.externalUrl}</a>.
  You cannot reply to this message.</small>
`;

export class EmailService extends AbstractService {
  public static create(config: Config): EmailService {
    const smtp = new SmtpClient(config);
    return new EmailService(config, smtp);
  }

  constructor(private config: Config, private smtp: SmtpClient) {
    super();
  }

  public confirmRegistration(lang: Language, to: string, token: string): Promise<void> {
    let subject: string;
    let content: string;
    const href = [this.config.externalUrl, Routes[lang].confirmAccount().format(), '?' + ConfirmAccountTokenParam + '=' + encodeURIComponent(token)].join('');

    switch (lang) {
      case Language.French:
        subject = 'Activation de votre compte Abc-Map';
        content = `
          <p>Bonjour !</p>
          <p>Pour activer votre compte Abc-Map, veuillez <a href="${href}" data-cy="enable-account-link">cliquer sur ce lien.</a></p>
          ${footerFr(this.config)}
        `;
        break;
      case Language.English:
        subject = 'Activation of your Abc-Map account';
        content = `
          <p>Hi !</p>
          <p>To activate your Abc-Map account, please <a href="${href}" data-cy="enable-account-link">click on this link.</a></p>
          ${footerEn(this.config)}
        `;
        break;
    }

    return this.smtp.sendMail(to, subject, content);
  }

  public resetPassword(lang: Language, to: string, token: string): Promise<void> {
    let subject: string;
    let content: string;

    const href = [this.config.externalUrl, Routes[lang].resetPassword().format(), '?' + ResetPasswordTokenParam + '=' + encodeURIComponent(token)].join('');

    switch (lang) {
      case Language.French:
        subject = 'Mot de passe perdu';
        content = `
          <p>Bonjour !</p>
          <p>Pour réinitialiser votre mot de passe Abc-Map, veuillez <a href="${href}" data-cy="reset-password-link">cliquer sur ce lien.</a></p>
          <p>Si vous n'êtes pas à l'origine de cette demande, ne tenez pas compte de ce message.</p>
          ${footerFr(this.config)}
        `;
        break;
      case Language.English:
        subject = 'Reset your password';
        content = `
          <p>Hi !</p>
          <p>To reset your Abc-Map password, please <a href="${href}" data-cy="reset-password-link">click on this link.</a></p>
          <p>If you are not the initiator of this request, ignore this message.</p>
          ${footerEn(this.config)}
        `;
        break;
    }
    return this.smtp.sendMail(to, subject, content);
  }
}
