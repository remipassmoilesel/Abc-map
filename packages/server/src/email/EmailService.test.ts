/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { EmailService } from './EmailService';
import { SmtpClient } from './SmtpClient';
import * as sinon from 'sinon';
import { SinonStubbedInstance } from 'sinon';
import { ConfigLoader } from '../config/ConfigLoader';
import { assert, expect } from 'chai';
import { Language } from '@abc-map/shared';

describe('EmailService', () => {
  let client: SinonStubbedInstance<SmtpClient>;
  let service: EmailService;

  beforeEach(async () => {
    const config = await ConfigLoader.load();
    client = sinon.createStubInstance(SmtpClient);
    service = new EmailService(config, client as unknown as SmtpClient);
  });

  describe('confirmRegistration()', () => {
    it('FR', async () => {
      await service.confirmRegistration(Language.French, 'to@domain', 'confirm-registration-token');

      assert.deepEqual(client.sendMail.callCount, 1);
      assert.deepEqual(client.sendMail.args[0][0], 'to@domain');
      assert.deepEqual(client.sendMail.args[0][1], 'Activation de votre compte Abc-Map');
      expect(client.sendMail.args[0][2]).toMatchSnapshot();
    });

    it('EN', async () => {
      await service.confirmRegistration(Language.English, 'to@domain', 'confirm-registration-token');

      assert.deepEqual(client.sendMail.callCount, 1);
      assert.deepEqual(client.sendMail.args[0][0], 'to@domain');
      assert.deepEqual(client.sendMail.args[0][1], 'Activation of your Abc-Map account');
      expect(client.sendMail.args[0][2]).toMatchSnapshot();
    });
  });

  describe('resetPassword()', () => {
    it('FR', async () => {
      await service.resetPassword(Language.French, 'to@domain', 'reset-password-token');

      assert.deepEqual(client.sendMail.callCount, 1);
      assert.deepEqual(client.sendMail.args[0][0], 'to@domain');
      assert.deepEqual(client.sendMail.args[0][1], 'Mot de passe perdu');
      expect(client.sendMail.args[0][2]).toMatchSnapshot();
    });

    it('EN', async () => {
      await service.resetPassword(Language.English, 'to@domain', 'reset-password-token');

      assert.deepEqual(client.sendMail.callCount, 1);
      assert.deepEqual(client.sendMail.args[0][0], 'to@domain');
      assert.deepEqual(client.sendMail.args[0][1], 'Reset your password');
      expect(client.sendMail.args[0][2]).toMatchSnapshot();
    });
  });
});
