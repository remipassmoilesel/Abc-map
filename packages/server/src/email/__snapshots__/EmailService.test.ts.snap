// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailService confirmRegistration() EN 1`] = `
"
          <p>Hi !</p>
          <p>To activate your Abc-Map account, please <a href=\\"http://localhost:10082/en/confirm-account/confirm-registration-token\\" data-cy=\\"enable-account-link\\">click on this link.</a></p>
          
  <p>&nbsp;</p>
  <small>This is an automatic message, sent by the platform <a href=\\"http://localhost:10082\\">http://localhost:10082</a>.
  You cannot reply to this message.</small>

        "
`;

exports[`EmailService confirmRegistration() FR 1`] = `
"
          <p>Bonjour !</p>
          <p>Pour activer votre compte Abc-Map, veuillez <a href=\\"http://localhost:10082/fr/confirm-account/confirm-registration-token\\" data-cy=\\"enable-account-link\\">cliquer sur ce lien.</a></p>
          
  <p>&nbsp;</p>
  <small>Ceci est un message automatique, envoyé par la plateforme <a href=\\"http://localhost:10082\\">http://localhost:10082</a>.
  Vous ne pouvez pas répondre à ce message.</small>

        "
`;

exports[`EmailService resetPassword() EN 1`] = `
"
          <p>Hi !</p>
          <p>To reset your Abc-Map password, please <a href=\\"http://localhost:10082/en/reset-password/reset-password-token\\" data-cy=\\"reset-password-link\\">click on this link.</a></p>
          <p>If you are not the initiator of this request, ignore this message.</p>
          
  <p>&nbsp;</p>
  <small>This is an automatic message, sent by the platform <a href=\\"http://localhost:10082\\">http://localhost:10082</a>.
  You cannot reply to this message.</small>

        "
`;

exports[`EmailService resetPassword() FR 1`] = `
"
          <p>Bonjour !</p>
          <p>Pour réinitialiser votre mot de passe Abc-Map, veuillez <a href=\\"http://localhost:10082/fr/reset-password/reset-password-token\\" data-cy=\\"reset-password-link\\">cliquer sur ce lien.</a></p>
          <p>Si vous n'êtes pas à l'origine de cette demande, ne tenez pas compte de ce message.</p>
          
  <p>&nbsp;</p>
  <small>Ceci est un message automatique, envoyé par la plateforme <a href=\\"http://localhost:10082\\">http://localhost:10082</a>.
  Vous ne pouvez pas répondre à ce message.</small>

        "
`;
